mission_control/include/mission_control/behavior.h

```diff
@@ -32,370 +32,65 @@
-
-  // Service-type behavior implementation overrides
-  virtual void callService() { }
-
-  // Common behavior overrides
-  virtual bool getParams(ros::NodeHandle nh) { return true; }
-  std::list<BehaviorXMLParam>& getBehaviorXmlParams() { return m_behaviorXMLParams; }
-  void setBehaviorXmlParams(std::list<BehaviorXMLParam> aList) { m_behaviorXMLParams = aList; }
-  virtual bool parseMissionFileParams() = 0;
-
-  TimeStamp getWhen() { return m_when; }
-  TimeStamp getTimeout() { return m_timeout; }
-  bool getTimeoutEna() { return m_timeout_ena; }
-  bool getBehaviorDone() { return m_behavior_done; }
-  float getBehaviorDuration() { return m_duration; }
-
-  int ExecuteBehavior(ros::NodeHandle nh)
-  {
-    switch (getType())
-    {
-      case BEHAVIOR_TYPE_MSG:
-        // Publish the appropriate behavior message
-        ROS_INFO("Publishing behavior [%s]", m_xml_tag.c_str());
-        publishMsg();
-        break;
-      case BEHAVIOR_TYPE_SRV:
-        ROS_INFO("Behavior [%s] calling service", m_xml_tag.c_str());
-        callService();
-        break;
-    }
-  }
-
-  int WaitForExecutionTimeSlot()
-  {
-    int retval = 0;  // 0 means success
-    boost::posix_time::ptime cur_ptime;
-    boost::posix_time::ptime calc_ptime;
-    cur_ptime = boost::posix_time::second_clock::universal_time();
-    TimeStamp when = getWhen();
-
-    switch (when.type)
-    {
-      case TimeStamp::REL_TIME:
-        if (when.duration.total_seconds() == 0) break;  // now
-        boost::this_thread::sleep(when.duration);
-        break;
-      case TimeStamp::ABS_TIME:
-        // Construct the appropriate absolute datetime
-        calc_ptime = boost::posix_time::ptime(cur_ptime.date(), when.time.time_of_day());
-        // Make sure to handle if the time is less than the current time (occurs next day)
-        if (calc_ptime.time_of_day() < cur_ptime.time_of_day())
-          calc_ptime += boost::gregorian::days(1);
-        boost::this_thread::sleep(boost::system_time(calc_ptime));
-        break;
-      case TimeStamp::ABS_DATETIME:
-        if (when.time < cur_ptime)
-        {
-          ROS_ERROR("Behavior occurs in the past; aborting mission");
-          retval = -1;
-          break;
-        }
-        boost::this_thread::sleep(boost::system_time(when.time));
-        break;
-    };
-    return retval;
-  }
+  ~Behavior() {}
 
-  int computeExecutionTimeForBehavior()
+  // You must override the virtual function tick()
+  BT::NodeStatus tick() override
   {
-    int nsec = -1;
-
-    TimeStamp tout = getTimeout();
-    boost::posix_time::ptime calc_ptime,
-        cur_ptime = boost::posix_time::second_clock::universal_time();
-
-    switch (tout.type)
+    switch (status())
     {
-      case TimeStamp::REL_TIME:
-        if (tout.duration.total_seconds() == 0) return 0;  // execution time is 0 seconds
-        nsec = tout.duration.total_seconds();
+      case BT::NodeStatus::IDLE:
+        setStatus(BT::NodeStatus::RUNNING);
```

Comment:
@patapufate why not doing any work on first tick?

Reply:
When we do the first tick, the status change from Idle to Running. Therefore, we need to check (in RUNNING State) that's the first tick.

Reply:
I won't block on it, but I'm not sure I understand how that explains why the first tick is effectively a no-op.